lab_1/main.py

```diff
@@ -3,81 +3,148 @@
+    freq_dict = calculate_frequencies(second_step)
+    if isinstance(language, str) and isinstance(text, str) and isinstance(stop_words, list):
+        language_profile = {"name": language,
+                            "freq": freq_dict,
+                            "n_words": len(freq_dict)}
+        return language_profile
+    return None
 
 
-def compare_profiles(unknown_profile: dict, profile_to_compare: dict, top_n: int) -> float or None:
+def compare_profiles(unknown_profile: dict, profile_to_compare: dict, top_n: int):
     """
     Compares profiles and calculates the distance using top n words
     :param unknown_profile: a dictionary
     :param profile_to_compare: a dictionary
     :param top_n: a number of the most common words
     :return: the distance
     """
-    pass
-
-
-def detect_language(unknown_profile: dict, profile_1: dict, profile_2: dict, top_n: int) -> str or None:
-    """
-    Detects the language of an unknown profile
-    :param unknown_profile: a dictionary
-    :param profile_1: a dictionary
-    :param profile_2: a dictionary
-    :param top_n: a number of the most common words
-    :return: a language
-    """
-    pass
-
-
-def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, top_n: int) -> list or None:
+    if isinstance(unknown_profile, dict) \
+            and isinstance(profile_to_compare, dict) \
+            and isinstance(top_n, int):
+        top_n_common_words = []
+        top_n_words_unknown_profile = get_top_n_words(unknown_profile['freq'], top_n)
+        top_n_words_known_profile = get_top_n_words(profile_to_compare['freq'], top_n)
+        for i in top_n_words_unknown_profile:
+            if i in top_n_words_known_profile:
+                top_n_common_words.append(i)
+        intersecting_words = len(top_n_common_words) / len(top_n_words_unknown_profile)
+        return round(float(intersecting_words), 2)
+    return None
+
+
+def detect_language(unknown_profile: dict, profile_1: dict, profile_2: dict, top_n: int):
+    """
+        Detects the language of an unknown profile
+        :param unknown_profile: a dictionary
+        :param profile_1: a dictionary
+        :param profile_2: a dictionary
+        :param top_n: a number of the most common words
+        :return: a language
+        """
+    if isinstance(unknown_profile, dict) \
+            and isinstance(profile_1, dict) \
+            and isinstance(profile_2, dict) \
+            and isinstance(top_n, int):
+        first_intersecting_words = compare_profiles(unknown_profile, profile_1, top_n)
+        second_intersecting_words = compare_profiles(unknown_profile, profile_2, top_n)
+        if first_intersecting_words > second_intersecting_words:
+            return profile_1['name']
+        if second_intersecting_words > first_intersecting_words:
+            return profile_2['name']
+        return [[profile_1['name'], profile_2['name']].sort()][0]
```

Comment:
can you describe in writing what happens here?

Reply:
first_intersecting_words and second_intersecting_words-float(proportion of intersecting words of the profile_1 and profile_2 that is returned by  compare_profiles) 
with their help we can detect the language:
-if the proportion of the language from first profile will be greater than from the second, the name of the first language returns and that will be the name of the unknown language
-however, if proportions are the same, names sorted alphabetically and the unkown language will be the language with index 0